README.md
# GUI Multimeter
This is a small multimeter that can be used to measure voltage and current in a circuit. The electronics are built using an Arduino Mega2560 (Arduino UNO will suffice as well), which can interface with both Ruby and Matlab code.

The project build updates are detailed in Changelog.md. Below are the instructions for building and running the multimeter.

<h3> The Circuit </h3>
The multimeter was built using a Funduino Mega2560, but can also be replicated with an Arduino Mega2560 or Arduino UNO. Any other microcontroller that supports 5V GPIO should work too.
To build the circuit, you will need: _a breadboard, four or five wires to go with it, and two resistors_. I used two 1.2 kΩ resistors, but any value between 100Ω and 10kΩ will work, as long as you adjust any calculations that require those resistance values.

The circuit diagram and pin connections to the Arduino can be found in this repository (circuit_diagram.png). First connect the negative rail on the breadboard to "ground" on the Arduino. Then put a wire into the analog input pin "A1" and connect it to a point on the breadboard. A single resistor should come both before and after the analog-in wire; by connecting a positive probe to the first resistor, a negative probe to the second resistor, and a wire to the negative rail after the second resistor, you will have created a voltage divider. Refer to the diagram for a better understanding of the circuit setup.

The arduino/breadboard combination will now accept up to 10V of input. **Important: Do not connect a voltage greater than 10V to the probes.** The Arduino Mega's analog-in pins are only designed to accept up to 5V. If you want to change the maximum voltage, you will have to choose a different resistor ratio, or wire more resistors up in series.

Before running either the Matlab or Ruby program, connect the 

<h3> Running the Voltmeter on Matlab (Windows or Linux) </h3>
Only the voltmeter has been implemented in Matlab. In order to run it, first make sure you have Matlab and the Arduino support package installed. Instructions here: http://www.allaboutcircuits.com/projects/arduino-interface-with-matlab/
Then, download the files "app\_layout.fig" and "app\_layout.m" from the repository. Make sure they are in your working directory, and run "app\_layout.m". The voltmeter should run automatically. Press "Measure" to start measuring the voltage between the two probes. Remember to connect the negative source to the negative probe and positive to positive!

<h3> Running the Multimeter on Ruby (Linux Recommended) </h3>
Requirements: Arduino IDE and latest Ruby version should be installed.

1. Install the Firmata gem. Instructions here: http://playground.arduino.cc/Interfacing/Ruby
2. Next, open Arduino IDE and run the "Standard Firmata" program with the Arduino connected to the computer. 
3. Install the gosu dependency by running `$ gem install gosu` on command line.
4. Download the files "connect.rb" and "voltm.rb" from the repository.
5. To test the arduino connection, run `$ ruby connect.rb`. If successful, the program will confirm the firmata connection and the version. Sometimes, it can take about one minute for a ruby program to actually detect the arduino. If the connection fails, press Ctrl+C to terminate the program, and try again until it connects.
6. You are ready to run the voltmeter now. Enter `$ ruby voltm.rb` in command line and a window should open up with a "measure" and "stop" button.
7. Press the "Measure" button to start measuring the voltage between two probes. If you want to measure current through the analog-input wire instead, press the "switch to amps" button below. You can switch between the voltmeter and ammeter at any time.

**Note**: The ammeter calculates the current passing through the point at which the analog-in wire intersects the circuit using the measured voltage. It assumes a voltage divider using two 1.2kΩ resistors. If you use a different resistor value, replace line 12 in voltm.rb that reads ` @@single_resistor = 1200 ` with your own resistance value in Ohms.
Additionally, by performing this calculation, _the ammeter ignores its own internal impedance_. As a result, the measurement will always be slightly inaccurate.


Changelog.md
# GUI Multimeter
This is a small project to construct a combined voltmeter and ammeter that will output results to a GUI. The GUI will be first built in Matlab and possibly C++. Diagrams may show any board, but all circuits and code files are being built around a Funduino Mega-2560.

Updates will be posted below, and this page will be replaced by a full README when the program is complete.

<h5> 2 August, 2016 </h5>
- Added initial code from Arduino IDE.
- Added circuit diagram

<h5> 6 August, 2016 </h5>
- Created a GUI in Matlab with start/stop button to use as counter.
- Adapted the voltage-measuring function from Arduino IDE to Matlab.

<h5> 7 August, 2016 </h5>
- Combined the voltage measuring function from 'voltm.m' into the app code in 'app_layout.m'.
- Voltmeter finished in debug mode: will try to run it through Matlab compiler later.

<h5> 11 August, 2016 </h5>
- Tweaked the circuit to make the wires a bit tidier: no change in sketch.
- Tried (and failed) to compile matlab GUI to executable.
- Uploaded screenshots of current view of app in three different circuit states.

<h5> 12 August, 2016 </h5>
- Began testing Arduino interfacing with Ruby. Instructions to run this test:
    First, install the "firmata" gem and arduino sketch using the instructions here: http://playground.arduino.cc/Interfacing/Ruby
    Next, download the file "Ruby Multimeter Testing.rb" and run it. The program should automatically detect your Arduino and connect to it.
- Current version is for command line only. Planning to add a GUI to it using Gosu.

<h5> 16 August, 2016 </h5>
- Finished a GUI prototype using Gosu. File has been added to the repository.
- Still some problems with timing and update cycles with the GUI. Once those are fixed, the command line voltmeter will be ready to be merged with the GUI.

<h5> 18 August, 2016 </h5>
- Combined the GUI with the initial voltmeter test: works fine except for a few things.
    1. The arduino arbitrarily fails to connect over firmata if it hasn't been plugged in long enough. Sometimes, many restarts are required. Once it connects, however, it stays put and does not disconnect.
    2. The Arduino's measurement cycles at 2 Hz are interfering with the GUI update cycles at 60 Hz. The plan to fix it is to keep both the updates at 60 Hz but create a timer so that the Arduino will only make a measurement every half-second.
- To run the Voltmeter GUI:
    First, install the "firmata" gem and run "standard firmata" through the Arduino IDE. Instructions here: http://playground.arduino.cc/Interfacing/Ruby . Then download the file "voltm.rb" and run it on command line. Press "Measure" to start measuring and *hold down* the stop button to stop measurement. This is because the GUI can only detect a mouse click every half-second due to the problem mentioned above. I'll be working to fix this shortly.

<h5> 22 August, 2016 </h5>
- Fixed "stop" button, and fixed firmata bugs: Arduino always connects now, and stop button always works (no need to hold it down anymore).
- Updates are back to 60 Hz: still trying to figure out a better way to solve that problem than using sleep(). Voltmeter is completely functional now. Set up the circuit, and run it using the instructions above. Adding an ammeter should now be a relatively trivial task.

<h5> 25 August, 2016 </h5>
- Added an ammeter, project now complete and functional
- Fixed circuit_setup PNG. New README on the way.


app_layout.m
function varargout = app_layout(varargin)
%APP_LAYOUT MATLAB code file for app_layout.fig
%      APP_LAYOUT, by itself, creates a new APP_LAYOUT or raises the existing
%      singleton*.
%
%      H = APP_LAYOUT returns the handle to a new APP_LAYOUT or the handle to
%      the existing singleton*.
%
%      APP_LAYOUT('Property','Value',...) creates a new APP_LAYOUT using the
%      given property value pairs. Unrecognized properties are passed via
%      varargin to app_layout_OpeningFcn.  This calling syntax produces a
%      warning when there is an existing singleton*.
%
%      APP_LAYOUT('CALLBACK') and APP_LAYOUT('CALLBACK',hObject,...) call the
%      local function named CALLBACK in APP_LAYOUT.M with the given input
%      arguments.
%
%      *See GUI Options on GUIDE's Tools menu.  Choose "GUI allows only one
%      instance to run (singleton)".
%
% See also: GUIDE, GUIDATA, GUIHANDLES

% Edit the above text to modify the response to help app_layout

% Last Modified by GUIDE v2.5 06-Aug-2016 14:50:14

% Begin initialization code - DO NOT EDIT
gui_Singleton = 1;
gui_State = struct('gui_Name',       mfilename, ...
                   'gui_Singleton',  gui_Singleton, ...
                   'gui_OpeningFcn', @app_layout_OpeningFcn, ...
                   'gui_OutputFcn',  @app_layout_OutputFcn, ...
                   'gui_LayoutFcn',  [], ...
                   'gui_Callback',   []);
if nargin && ischar(varargin{1})
   gui_State.gui_Callback = str2func(varargin{1});
end

if nargout
    [varargout{1:nargout}] = gui_mainfcn(gui_State, varargin{:});
else
    gui_mainfcn(gui_State, varargin{:});
end
% End initialization code - DO NOT EDIT


% --- Executes just before app_layout is made visible.
function app_layout_OpeningFcn(hObject, eventdata, handles, varargin)
% This function has no output args, see OutputFcn.
% hObject    handle to figure
% eventdata  reserved - to be defined in a future version of MATLAB
% handles    structure with handles and user data (see GUIDATA)
% varargin   unrecognized PropertyName/PropertyValue pairs from the
%            command line (see VARARGIN)

handles.call_break=5;
% Choose default command line output for app_layout
handles.output = hObject;
set(handles.text4,'String','Press Measure to Start');
% Update handles structure
guidata(hObject, handles);

% UIWAIT makes app_layout wait for user response (see UIRESUME)
% uiwait(handles.figure1);


% --- Outputs from this function are returned to the command line.
function varargout = app_layout_OutputFcn(hObject, eventdata, handles)
% varargout  cell array for returning output args (see VARARGOUT);
% hObject    handle to figure
% eventdata  reserved - to be defined in a future version of MATLAB
% handles    structure with handles and user data (see GUIDATA)

% Get default command line output from handles structure
varargout{1} = handles.output;


% --- Executes on button press in stop_btn.
function stop_btn_Callback(hObject, eventdata, handles)
% hObject    handle to stop_btn (see GCBO)
% eventdata  reserved - to be defined in a future version of MATLAB
% handles    structure with handles and user data (see GUIDATA)
handles.call_break=1;
guidata(hObject,handles);
set(handles.text4,'String','Press Measure to Start');
drawnow;


% --- Executes on button press in msr_btn.
function msr_btn_Callback(hObject, eventdata, handles)
% hObject    handle to msr_btn (see GCBO)
% eventdata  reserved - to be defined in a future version of MATLAB
% handles    structure with handles and user data (see GUIDATA)
handles.call_break=0;
guidata(hObject,handles);
clear ard;
ard=arduino('COM3','Mega2560');
while(true)
    redval = readVoltage(ard,'A1');
    volt_str=sprintf('%.2f V',redval*2);
    set(handles.text4,'String',volt_str);
    pause(1);
    handles=guidata(hObject);
    if(handles.call_break==1)
        break;
    end
end


voltmeter-gosu-GUI-only.rb
 require 'rubygems'
 require 'gosu'

 module Zorder
 	Background, Buttons, Text = *0..2;
 end

 class MeterWindow < Gosu::Window
 	def initialize
 		puts Zorder::Background
 		@first_draw=1;
 		@width_wrapper=720;
 		@height_wrapper=480;
 		@background_color=Gosu::Color.argb(0xff_808080);
 		@redr=true;
 		@font_large=Gosu::Font.new(72,{:name => "Lato Medium"});
 		@font_medium=Gosu::Font.new(45,{:name => "Lato Medium"})
 		@font_small=Gosu::Font.new(36,{:name => "Ubuntu"});
 		@measuring = false;
 		needs_cursor=true;
 		super(720,480);
 		self.caption = "Voltmeter test";
 	end

 	def update
 		@redr=false;
 		@msx=mouse_x;
 		@msy=mouse_y;
 		@measuring = true if measure_clicked?
 		@measuring = false if stop_clicked?
 		@first_draw = 1 if stop_clicked?
 		if @measuring then @redr=true else @redr=false end;
 	end

 	def measure_clicked?
 		if button_down?(Gosu::MsLeft)
 			if (@msx.between?(100,250) and @msy.between?(110,190))
 				return true;
 			end
 		end
 		return false;
 	end

 	def stop_clicked?
 		if button_down?(Gosu::MsLeft)
 			if (@msx.between?(489,639) and @msy.between?(110,190))
 				return true;
 			end
 		end
 		return false;
 	end

 	def needs_cursor?
 		true;
 	end

 	def needs_redraw?
 		if @first_draw==1
 			return true;
 		else
 			return @redr;
 		end
 	end

 	def draw
 		# puts "Redrawing..."; # use for debugging drawing updates
 		draw_rect(0,0,@width_wrapper,@height_wrapper, \
 			      @background_color,Zorder::Background);
 		@font_large.draw("VOLTAGE",257,30,Zorder::Text,1,1,0xff_00FF44);

 		# Measure and Stop buttons
 		draw_rect(100,110,150,80,0xff_AAAAAA,Zorder::Buttons);
 		@font_small.draw("MEASURE",118,132,Zorder::Text,1,1,0xff_000000);
 		draw_rect(489,110,150,80,0xff_AAAAAA,Zorder::Buttons);
 		@font_small.draw("STOP",530,132,Zorder::Text,1,1,0xff_000000);

 		# Voltage reader
 		stopped_str="Press \"Measure\" to start";
 		unless @measuring
 			@font_medium.draw(stopped_str,210,290,
 				              Zorder::Text,1,1,0xff_00FF44); 		
 		else
 			val=rand(1000);
 			@font_medium.draw("#{val} V",320,290,
 				              Zorder::Text,1,1,0xff_00FF44);
 			sleep(0.5);
 		end
 		@first_draw=0;
 	end

 	def draw_rect(x_in,y_in,width_in,height_in,color_in,z)
 		draw_quad(x_in,y_in,color_in, \
 			 	  x_in+width_in,y_in,color_in, \
 			 	  x_in+width_in,y_in+height_in,color_in, \
 			 	  x_in,y_in+height_in,color_in,z);
 	end

 end

 Window=MeterWindow.new;
 Window.show;

voltm.rb
 require 'gosu'
 require 'arduino_firmata'


 module Zorder
 	Background, Buttons, Text = *0..2;
 end

 class MeterWindow < Gosu::Window
 	def initialize
 		
 		@@single_resistor = 1200 # The resistance of ONE resistor in the voltage divider

 		@@unit_mode_strings=["Volts","Amps"];
 		@first_draw=1;
 		@width_wrapper=720;
 		@height_wrapper=480;
 		@background_color=Gosu::Color.argb(0xff_808080);
 		@redr=true;
 		@font_large=Gosu::Font.new(72,{:name => "Lato Medium"});
 		@font_medium=Gosu::Font.new(45,{:name => "Lato Medium"})
 		@font_small=Gosu::Font.new(36,{:name => "Ubuntu"});
 		@font_tiny=Gosu::Font.new(24,{:name => "Ubuntu"});
 		@measuring = false;
 		@counter=0;
 		@val = 0.00;
 		@unit_mode=1; # 1 for volts, 2 for amps 
 		needs_cursor=true;
 		super(720,480);
 		self.caption = "Voltmeter test";

 		print "Connecting..."
 		@ard = ArduinoFirmata.connect
 		puts "Connection successful: Firmata version #{@ard.version}"

 	end

 	def update
 		@redr=false;
 		@msx=mouse_x;
 		@msy=mouse_y;
 		@measuring = true if measure_clicked?
 		if stop_clicked?
 			@measuring=false;
 			@first_draw=1;
 			@counter=0;
 		end

 		if switch_clicked?
 			@unit_mode=3-@unit_mode;
 			sleep(0.5);
 		end

 		if @measuring
 			@redr=true
 			@counter = (@counter + 1)%30;
 		else
 			@redr=false 
 		end;
 	end

 	def measure_clicked?
 		if button_down?(Gosu::MsLeft)
 			if (@msx.between?(100,250) and @msy.between?(110,190))
 				return true;
 			end
 		end
 		return false;
 	end

 	def stop_clicked?
 		if button_down?(Gosu::MsLeft)
 			if (@msx.between?(489,639) and @msy.between?(110,190))
 				return true;
 			end
 		end
 		return false;
 	end

 	def switch_clicked?
 		if @measuring and button_down?(Gosu::MsLeft)
 			if(@msx.between?(275,465) and @msy.between?(380,415))
 				return true;
 			end
 		end
 		return false;
 	end

 	def needs_cursor?
 		true
 	end

 	def needs_redraw?
 		if @first_draw==1
 			return true
 		else
 			return @redr
 		end
 	end

 	def draw
 		# puts "Redrawing..."; # use for debugging drawing updates
 		draw_rect(0,0,@width_wrapper,@height_wrapper, \
 			      @background_color,Zorder::Background);
 		@font_large.draw("VOLTAGE",257,30,Zorder::Text,1,1,0xff_00FF44);

 		# Measure and Stop buttons
 		draw_rect(100,110,150,80,0xff_AAAAAA,Zorder::Buttons);
 		@font_small.draw("MEASURE",118,132,Zorder::Text,1,1,0xff_000000);
 		draw_rect(489,110,150,80,0xff_AAAAAA,Zorder::Buttons);
 		@font_small.draw("STOP",530,132,Zorder::Text,1,1,0xff_000000);
 		#Mode switch button
 		if @measuring
	 		draw_rect(275,380,190,35,0xff_FF889B,Zorder::Buttons);
	 		@font_tiny.draw("Switch to #{@@unit_mode_strings[2-@unit_mode]}", \
	 			             304,388,Zorder::Text,1,1,0xff_000000);
 		end

 		# Voltage reader
 		stopped_str="Press \"Measure\" to start";
 		unless @measuring
 			@font_medium.draw(stopped_str,210,290,
 				              Zorder::Text,1,1,0xff_00FF44); 		
 		else
 			# The final value of "val" is the voltage between ground
 			# and one resistor.  Current will be measured at the point
 			# at which the analog-in probe is connected to any circuit.
 			if(@counter==0)
 				@val = @ard.analog_read 1;
    			@val = @val * (5.0/1024.0) * 2;
    		end
    		if @unit_mode==1 # measuring volts
    			val_str=sprintf("%.2f",@val) + " V";
    		elsif @unit_mode==2 # measuring amps
    			curr=@val/(2*@@single_resistor) * 1000;
    			val_str=sprintf("%.2f",curr) + " mA";
    		end
 			@font_medium.draw(val_str,320,290,
 				              Zorder::Text,1,1,0xff_00FF44);
 			# sleep(0.5);
 		end
 		@first_draw=0;
 	end

 	def draw_rect(x_in,y_in,width_in,height_in,color_in,z)
 		draw_quad(x_in,y_in,color_in, \
 			 	  x_in+width_in,y_in,color_in, \
 			 	  x_in+width_in,y_in+height_in,color_in, \
 			 	  x_in,y_in+height_in,color_in,z);
 	end

 end

 Window=MeterWindow.new;
 Window.show;